packages/ipns/src/utils/dns.ts

```diff
@@ -0,0 +1,99 @@
+import { CodeError } from '@libp2p/interfaces/errors'
+import * as isIPFS from 'is-ipfs'
+import type { ResolveDnsLinkOptions } from '../index.js'
+
+export interface Question {
+  name: string
+  type: number
+}
+
+export interface Answer {
+  name: string
+  type: number
+  TTL: number
+  data: string
+}
+
+export interface DNSResponse {
+  Status: number
+  TC: boolean
+  RD: boolean
+  RA: boolean
+  AD: boolean
+  CD: boolean
+  Question: Question[]
+  Answer: Answer[]
+}
+
+export const ipfsPath = (domain: string, response: DNSResponse): string => {
+  const answer = findDNSLinkAnswer(domain, response)
+
+  let data = answer.data
+
+  if (data.startsWith('"')) {
+    data = data.substring(1)
+  }
+
+  if (data.endsWith('"')) {
+    data = data.substring(0, data.length - 1)
+  }
+
+  return data.replace('dnslink=', '')
+}
+
+export const findDNSLinkAnswer = (domain: string, response: DNSResponse): Answer => {
+  const answer = response.Answer.filter(a => a.data.includes('dnslink=/ipfs') || a.data.includes('dnslink=/ipns')).pop()
+
+  if (answer == null) {
+    throw new CodeError(`No dnslink records found for domain: ${domain}`, 'ERR_DNSLINK_NOT_FOUND')
+  }
+
+  return answer
+}
+
+export const findTTL = (domain: string, response: DNSResponse): number => {
+  const answer = findDNSLinkAnswer(domain, response)
+
+  return answer.TTL
+}
+
+export const MAX_RECURSIVE_DEPTH = 32
+
+export const recursiveResolveDnslink = async (domain: string, depth: number, resolve: (domain: string, options: ResolveDnsLinkOptions) => Promise<string>, options: ResolveDnsLinkOptions = {}): Promise<string> => {
+  if (depth === 0) {
+    throw new Error('recursion limit exceeded')
+  }
+
+  let dnslinkRecord: string
+
+  try {
+    dnslinkRecord = await resolve(domain, options)
+  } catch (err: any) {
+    // If the code is not ENOTFOUND or ERR_DNSLINK_NOT_FOUND or ENODATA then throw the error
+    if (err.code !== 'ENOTFOUND' && err.code !== 'ERR_DNSLINK_NOT_FOUND' && err.code !== 'ENODATA') {
+      throw err
+    }
+
+    if (domain.startsWith('_dnslink.')) {
+      // The supplied domain contains a _dnslink component
+      // Check the non-_dnslink domain
+      domain = domain.replace('_dnslink.', '')
+    } else {
+      // Check the _dnslink subdomain
+      domain = `_dnslink.${domain}`
+    }
+
+    // If this throws then we propagate the error
+    dnslinkRecord = await resolve(domain, options)
+  }
+
+  const result = dnslinkRecord.replace('dnslink=', '')
```

Comment:
technically, depth inside this scope was already checked, and the `depth===0` at the top would have checked if it was 0. The check below on line 94 is a duplicate and should never fire. So we should technically check if depth is 1, and we can return if we want to prevent erroring on further recursion to allow a user to do so if they wish?

@achingbrain what do you think

Reply:
it really depends on whether we want to allow infinite recursion and error if infinite recursion exists, or recurse as far as MAX and return what we have so far (user could then request to resolve that further if they want)

could be a config option.